File: booking/notifier.py
"""Отправка уведомлений о бронировании по электронной почте.

Используется только стандартная библиотека: :mod:`smtplib` и
:mod:`email.message`. Если SMTP-сервер не настроен, письмо печатается в
консоль — это позволяет проверить работу команды без реального сервера.
"""

from __future__ import annotations

import smtplib
import ssl
from email.headerregistry import Address
from email.message import EmailMessage
from typing import Protocol

from booking.config import SmtpConfig
from booking.errors import NotificationError
from booking.models import Booking, Room

_SUBJECT_TEMPLATE = "Бронь кабинета №{room_id} на {date} {time}"

_BODY_TEMPLATE = """Здравствуйте, {name}!

Кабинет забронирован за вами.

    Кабинет:      №{room_id} — {room_name}
    Дата:         {date}
    Время:        {time_range} ({duration} мин.)
    Забронировал: {name} <{email}>{purpose_line}

Номер брони: {booking_id}
Чтобы отменить бронь, выполните: python -m booking cancel {booking_id}

Это письмо сформировано автоматически, отвечать на него не нужно.
"""


def build_message(
    booking: Booking, room: Room, config: SmtpConfig
) -> EmailMessage:
    """Сформировать письмо с датой, временем и номером кабинета."""
    purpose_line = (
        f"\n    Тема:         {booking.purpose}" if booking.purpose else ""
    )
    message = EmailMessage()
    message["Subject"] = _SUBJECT_TEMPLATE.format(
        room_id=room.id,
        date=f"{booking.slot.start:%d.%m.%Y}",
        time=booking.slot.human_time(),
    )
    sender_user, _, sender_domain = config.sender.partition("@")
    message["From"] = Address(
        config.sender_name, sender_user, sender_domain
    )
    message["To"] = booking.person_email
    message.set_content(
        _BODY_TEMPLATE.format(
            name=booking.person_name,
            email=booking.person_email,
            room_id=room.id,
            room_name=room.name,
            date=f"{booking.slot.start:%d.%m.%Y}",
            time_range=booking.slot.human_time(),
            duration=booking.slot.duration_minutes,
            purpose_line=purpose_line,
            booking_id=booking.id,
        )
    )
    return message


class Notifier(Protocol):
    """Интерфейс отправителя уведомлений."""

    def send(self, message: EmailMessage) -> str:
        """Отправить письмо и вернуть описание результата."""
        ...


class ConsoleNotifier:
    """Заглушка: печатает письмо в консоль вместо отправки.

    Используется, когда SMTP не настроен, а также в режиме ``--dry-run``.
    """

    def send(self, message: EmailMessage) -> str:
        """Вывести письмо в стандартный поток вывода."""
        print("-" * 62)
        print(f"Кому:  {message['To']}")
        print(f"Тема:  {message['Subject']}")
        print("-" * 62)
        print(message.get_content().rstrip())
        print("-" * 62)
        return (
            f"письмо выведено в консоль (SMTP не настроен), "
            f"адресат: {message['To']}"
        )


class SmtpNotifier:
    """Отправка письма через реальный SMTP-сервер."""

    def __init__(self, config: SmtpConfig) -> None:
        """Сохранить параметры подключения."""
        self._config = config

    def send(self, message: EmailMessage) -> str:
        """Отправить письмо, преобразуя ошибки SMTP в доменные."""
        config = self._config
        try:
            if config.port == 465:
                server = smtplib.SMTP_SSL(
                    config.host,
                    config.port,
                    timeout=30,
                    context=ssl.create_default_context(),
                )
            else:
                server = smtplib.SMTP(config.host, config.port, timeout=30)
            with server:
                server.ehlo()
                if config.use_tls and config.port != 465:
                    server.starttls(context=ssl.create_default_context())
                    server.ehlo()
                if config.user:
                    server.login(config.user, config.password)
                server.send_message(message)
        except (smtplib.SMTPException, OSError) as exc:
            raise NotificationError(
                f"Не удалось отправить письмо на {message['To']}: {exc}"
            ) from exc
        return f"письмо отправлено на {message['To']}"


def get_notifier(config: SmtpConfig, dry_run: bool = False) -> Notifier:
    """Выбрать отправителя в зависимости от конфигурации."""
    if dry_run or not config.is_configured:
        return ConsoleNotifier()
    return SmtpNotifier(config)

File: internal/booking/notifier.go
package booking

import (
	"crypto/tls"
	"fmt"
	"io"
	"mime"
	"net"
	"net/smtp"
	"os"
	"strings"
)

// Message — письмо, готовое к отправке.
type Message struct {
	To      string
	Subject string
	Body    string
	from    string
}

// bodyTemplate — текст уведомления о брони.
const bodyTemplate = `Здравствуйте, %s!

Кабинет забронирован за вами.

    Кабинет:      №%d — %s
    Дата:         %s
    Время:        %s (%d мин.)
    Забронировал: %s <%s>%s

Номер брони: %d
Чтобы отменить бронь, выполните: booking cancel %d

Это письмо сформировано автоматически, отвечать на него не нужно.
`

// BuildMessage формирует письмо с датой, временем и номером кабинета.
func BuildMessage(b Booking, room Room, config SMTPConfig) Message {
	purposeLine := ""
	if b.Purpose != "" {
		purposeLine = "\n    Тема:         " + b.Purpose
	}
	date := b.Slot.Start.Format("02.01.2006")
	return Message{
		To: b.PersonEmail,
		Subject: fmt.Sprintf("Бронь кабинета №%d на %s %s",
			room.ID, date, b.Slot.HumanTime()),
		Body: fmt.Sprintf(bodyTemplate, b.PersonName, room.ID, room.Name,
			date, b.Slot.HumanTime(), b.Slot.Minutes(),
			b.PersonName, b.PersonEmail, purposeLine, b.ID, b.ID),
		from: fmt.Sprintf("%s <%s>",
			mime.QEncoding.Encode("utf-8", config.SenderName), config.Sender),
	}
}

// Bytes возвращает письмо в виде RFC 5322 с заголовками в UTF-8.
func (m Message) Bytes() []byte {
	var b strings.Builder
	fmt.Fprintf(&b, "From: %s\r\n", m.from)
	fmt.Fprintf(&b, "To: %s\r\n", m.To)
	fmt.Fprintf(&b, "Subject: %s\r\n",
		mime.QEncoding.Encode("utf-8", m.Subject))
	b.WriteString("MIME-Version: 1.0\r\n")
	b.WriteString("Content-Type: text/plain; charset=UTF-8\r\n")
	b.WriteString("Content-Transfer-Encoding: 8bit\r\n\r\n")
	b.WriteString(strings.ReplaceAll(m.Body, "\n", "\r\n"))
	return []byte(b.String())
}

// Notifier отправляет уведомления.
type Notifier interface {
	Send(Message) (string, error)
}

// NewNotifier выбирает отправителя в зависимости от конфигурации.
//
// Пока SMTP не настроен, письма печатаются в out: утилита остаётся
// полностью работоспособной без почтового сервера. Поток передаётся
// параметром, а не берётся из os.Stdout, иначе вывод нельзя было бы
// перенаправить — ни в тестах, ни при запуске с другими потоками.
func NewNotifier(config SMTPConfig, dryRun bool, out io.Writer) Notifier {
	if out == nil {
		out = os.Stdout
	}
	if dryRun || !config.IsConfigured() {
		return ConsoleNotifier{Out: out}
	}
	return SMTPNotifier{config: config}
}

// ConsoleNotifier печатает письмо вместо отправки.
type ConsoleNotifier struct {
	Out io.Writer
}

// Send выводит письмо в заданный поток.
func (n ConsoleNotifier) Send(m Message) (string, error) {
	line := strings.Repeat("-", 62)
	fmt.Fprintf(n.Out, "%s\nКому:  %s\nТема:  %s\n%s\n%s%s\n",
		line, m.To, m.Subject, line, m.Body, line)
	return fmt.Sprintf(
		"письмо выведено в консоль (SMTP не настроен), адресат: %s", m.To), nil
}

// SMTPNotifier отправляет письмо через реальный почтовый сервер.
type SMTPNotifier struct {
	config SMTPConfig
}

// Send отправляет письмо, оборачивая ошибки транспорта в ErrNotification.
func (n SMTPNotifier) Send(m Message) (string, error) {
	if err := n.send(m); err != nil {
		return "", fmt.Errorf("%w: письмо на %s: %v", ErrNotification, m.To, err)
	}
	return fmt.Sprintf("письмо отправлено на %s", m.To), nil
}

// send выполняет сеанс SMTP: порт 465 — сразу TLS, иначе STARTTLS.
func (n SMTPNotifier) send(m Message) error {
	addr := net.JoinHostPort(n.config.Host, fmt.Sprint(n.config.Port))

	var client *smtp.Client
	var err error
	if n.config.Port == 465 {
		conn, dialErr := tls.Dial("tcp", addr,
			&tls.Config{ServerName: n.config.Host})
		if dialErr != nil {
			return dialErr
		}
		if client, err = smtp.NewClient(conn, n.config.Host); err != nil {
			return err
		}
	} else {
		if client, err = smtp.Dial(addr); err != nil {
			return err
		}
		if n.config.UseTLS {
			if err = client.StartTLS(
				&tls.Config{ServerName: n.config.Host}); err != nil {
				return err
			}
		}
	}
	defer client.Quit()

	if n.config.User != "" {
		auth := smtp.PlainAuth("", n.config.User, n.config.Password, n.config.Host)
		if err = client.Auth(auth); err != nil {
			return err
		}
	}
	if err = client.Mail(n.config.Sender); err != nil {
		return err
	}
	if err = client.Rcpt(m.To); err != nil {
		return err
	}
	writer, err := client.Data()
	if err != nil {
		return err
	}
	if _, err = writer.Write(m.Bytes()); err != nil {
		return err
	}
	return writer.Close()
}
